libbpf-tools/tcpretrans.c

```diff
@@ -0,0 +1,359 @@
+		warn("dump_hash: %s", strerror(errno));
+		return;
+	}
+
+	for (i = 0; i < n; i++) {
+		src.s_addr = keys[i].saddr;
+		dst.s_addr = keys[i].daddr;
+		sprintf(local, ep_fmt, inet_ntop(AF_INET, &src, s, sizeof(s)), keys[i].sport);
+		sprintf(remote, ep_fmt, inet_ntop(AF_INET, &dst, d, sizeof(d)), ntohs(keys[i].dport));
+		printf("%-20s <-> %-20s %10llu\n", local, remote, counts[i]);
+	}
+}
+
+static void print_count_ipv6(int map_fd)
+{
+	static struct ipv6_flow_key keys[MAX_ENTRIES];
+	__u32 value_size = sizeof(__u64);
+	__u32 key_size = sizeof(keys[0]);
+	static struct ipv6_flow_key zero;
+	static __u64 counts[MAX_ENTRIES];
+	char s[INET6_ADDRSTRLEN];
+	char d[INET6_ADDRSTRLEN];
+	__u32 i, n = MAX_ENTRIES;
+	struct in6_addr src;
+	struct in6_addr dst;
+	char *ep_fmt = "[%s]#%d";
+	char remote[INET6_ADDRSTRLEN + 8];
+	char local[INET6_ADDRSTRLEN + 8];
+
+	if (dump_hash(map_fd, keys, key_size, counts, value_size, &n, &zero)) {
+		warn("dump_hash: %s", strerror(errno));
+		return;
+	}
+
+	for (i = 0; i < n; i++) {
+		memcpy(src.s6_addr, keys[i].saddr, sizeof(src.s6_addr));
+		memcpy(dst.s6_addr, keys[i].daddr, sizeof(src.s6_addr));
+		sprintf(local, ep_fmt, inet_ntop(AF_INET6, &src, s, sizeof(s)), keys[i].sport);
+		sprintf(remote, ep_fmt, inet_ntop(AF_INET6, &dst, d, sizeof(d)), ntohs(keys[i].dport));
+		printf("%-20s <-> %-20s %10llu\n", local, remote, counts[i]);
+	}
+}
+
+static void print_count(int map_fd_ipv4, int map_fd_ipv6)
+{
+	while (hang_on)
+		pause();
+
+	printf("\n%-25s %-25s %-10s\n", "LADDR:LPORT", "RADDR:RPORT", "RETRANSMITS");
+	print_count_ipv4(map_fd_ipv4);
+	print_count_ipv6(map_fd_ipv6);
+}
+
+static void print_events_header()
+{
+	printf("%-8s %-6s %-2s %-20s %1s> %-20s %-4s\n", "TIME", "PID", "IP",
+		"LADDR:LPORT", "T", "RADDR:RPORT", "STATE");
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	char src[INET6_ADDRSTRLEN];
+	char dst[INET6_ADDRSTRLEN];
+	char remote[INET6_ADDRSTRLEN + 6];
+	char local[INET6_ADDRSTRLEN + 6];
+	union {
+		struct in_addr  x4;
+		struct in6_addr x6;
+	} s, d;
+
+	if (e->af == AF_INET) {
+		memcpy(&s.x4.s_addr, e->saddr, sizeof(s.x4.s_addr));
+		memcpy(&d.x4.s_addr, e->daddr, sizeof(d.x4.s_addr));
+	} else if (e->af == AF_INET6) {
+		memcpy(&s.x6.s6_addr, e->saddr, sizeof(s.x6.s6_addr));
+		memcpy(&d.x6.s6_addr, e->daddr, sizeof(d.x6.s6_addr));
+	} else {
+		warn("broken event: event->af=%d", e->af);
+		return;
+	}
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	sprintf(local, "%s:%d", inet_ntop(e->af, &s, src, sizeof(src)), e->sport);
+	sprintf(remote, "%s:%d", inet_ntop(e->af, &d, dst, sizeof(dst)), ntohs(e->dport));
+
+	printf("%-8s %-6d %-2d %-20s %1s> %-20s %s\n",
+		   ts,
+		   e->pid,
+		   e->af == AF_INET ? 4 : 6,
+		   local,
+		   e->type == RETRANSMIT ? "R" : "L",
+		   remote,
+		   TCPSTATE[e->state - 1]);
```

Comment:
if e->state somehow becomes 0 (e.g., failed read or something), you'll be reading way outside of array. Maybe let's add entry 0 in TCPSTATE like "<INVALID>"? And then also check that e->state < ARRAY_SIZE(TCPSTATE)?